### DISCUSSION TITLE
ARM support / Progress with Surface Pro X

### DISCUSSION THREAD

**Discussion by stefanschmitt** [linux-surface/linux-surface#636] [General]
I am able to boot Debian and Arch Linux Kernel 5.14/5.15 with ACPI (not DT) on a Surface Pro X SQ1

- PCIe NVME shows up in lspci but no device file for access / external drive works
- Internal USB3 and Surface Dock 2 USB3 ports as well as Ethernet works (via Surface Connect)
- Signature Keyboard is not recognized as far as I can see / external kb works

Feels great to browse this repo using Elinks on SPX. How can I help any further without skills in hardware programming? How to compile Surface-Linux tools for ARM? Does this make any sense at all? 

Thanks in advance 
Stefan

**Comment by qzed** [comment]
Well yeah, that's a bit of a mixed bag. For now, I don't think it'll be particularly usable.

There's a fix to get the NVME working with ACPI (https://github.com/linux-surface/kernel/commit/9e032573a61358925c78b95f77f0ea9ec7e7cfc9) and there are also some pretty ugly hacks to get the keyboard working (in particular https://github.com/linux-surface/kernel/commit/1dfde018f9460cfa497108ced260268c0fdf5a87, which is required to not exit UEFI mode as that would un-configure certain clocks and subsystems required for the keyboard; something I don't particularly recommend). The way the SPX (and other Qualcomm Windows devices) are set up is that a lot of information about the SoC and how it needs to be configured is actually stored in the drivers. Meaning we'll probably have to reverse engineer those to get what we need (or prevent UEFI from un-configuring devices as in the hacks, but that will only get us so far, as in only get us the devices in those specific modes they are set up/needed in UEFI). An ACPI based working-tree is here: https://github.com/linux-surface/kernel/commits/v5.12-spx (note that keyboard doesn't work on that because I've reverted the UEFI hack there, so you'll have to revert the [revert](https://github.com/linux-surface/kernel/commit/833b9c110ffd41019aa87fe4bf829ef593605822) if you want that working).

IMHO a more promising approach is using the drivers that (among others) Bjorn Andersson of Linaro is (has been?) working on for the Lenovo Flex 5G (same SC8180 SoC). Those are, however, (like pretty much everything with Linux on ARM) DT-based. I've tried working off of his work and creating a DTS file for the SPX, but on that the USB for the dock isn't working (so no dock at all, including ethernet, as that's all USB-based) and the NVME doesn't work either (also no WiFi). But the keyboard cover works. You can find a tree for that here: https://github.com/linux-surface/kernel/tree/v5.12-spx-dt. Most of the missing stuff is blocked by lack of information (at least as far as I can tell): It should be possible to at least get the NVME and WiFi working by only modifying the DTS. Probably GPU support and all the other stuff that's working on the Flex 5G as well. IIRC the USB for the dock might need some more work, but I can't really tell since I'm pretty much a beginner with the whole DT/SoC-level stuff myself. 

I don't have the time right now to work on this (and probably for the next quarter to half year, hopefully after that I can dig into this some more). See also https://github.com/Sonicadvance1/linux/issues/27.

**Reply by philliplanos** [reply]
Anyone who gets a full running Linux on the SPX will be doing gods work, man.

**Reply by qzed** [reply]
We now have a repo for the SPX and its status: https://github.com/linux-surface/surface-pro-x/.
